### DISCUSSION TITLE
RLS issues in Supabase arise when attempting to filter products based on the `company_id` stored in `auth.user_metadata`

### DISCUSSION THREAD

**Discussion by ivanasetiawan** [supabase/supabase#20373] [Questions] [answered]
I want to show a list of products to the right employees. Based on the structure below:
- Tina works at Company A (10), so I want them to see Apple
- Sam works at Company B (11), so they will see Banana.

It sounds so straightforward, but somehow, RLS behaved unpredictably when I created a condition based on `auth.jwt()`. 

Here are my tables:
Company table
| id |    name   |
|----|:---------:|
| **10** | Company A |
| **11** | Company B |

User table
| id    | company_id |  name |
|-------|:----------:|------:|
| uuid1 |  **10**[FK] (to company table) | Tina  |
| uuid2 |     **11**[FK] (to company table)    | Sam   |

Product table
| id |    name   |  company_id |
|----|:---------:|:-----------:|
| 1  | Apple     | **10**[FK] (to company table) |
| 2  | Banana    | **11**[FK] (to company table) |

I have a function that is triggered to set `auth.usermeta` & it works as it should:
```
const { data:authData, error: authErr } = await supabase.auth.updateUser({
    data: {
        company_id, // company_id is a string I got from the session
         role: 'manager'
    }
});
```

Now on the product table on Supabase, I set this RLS on SELECT:
```
auth.jwt() -> 'user_metadata' ->> 'company_id' = (company_id)::text
```

Everything seems to be going OK (no error), however - I see the wrong products (and sometimes even no product) on user's dashboard. I have tested different SQL code including converting `user_metadata` `company_id` to `bigint`:
```
((((auth.jwt() -> 'user_metadata'::text) ->> 'company_id'::text))::bigint = company_id)
```

But same result, products just don't seem to show correctly :(

I finally gave up tackle this from the FE and it works as expected:
```
const { data, error: err } = await supabase
    .from('product')
    .select()
    .eq('company_id', session?.user.user_metadata.company_id);
```

I prefer to solve this using RLS instead of always passing `session?.user.user_metadata.company_id`. Why don't the products show correctly when I used the SQL expression?

_My project uses Supabase(^2.39.1) and Sveltekit(^2.0.0)_

**Comment by GaryAustin1** [comment]
I'm going to guess your changes to auth user metadata from the updateUser() are not reflecting in the user jwt yet.  Changing the auth table does not change the jwt unless the jwt refreshes when it expires or you do something like getUser() that causes the jwt to refresh with the latest info.

**Reply by ivanasetiawan** [reply]
How do I ensure I have the most up-to-date user information in Supabase? I checked `auth.users` and `user_metadata` is present with the right "latest" `company_id`; also logged the session, and it shows the correct info. However, you mentioned it might not be the latest version... Does this mean the frontend relies on `auth.users` while SQL uses the potentially outdated JWT info? Do I need to execute on the FE => `updateUser`, then `getUser` to refresh the JWT, and finally run the SQL query?

Thank you in advance @GaryAustin1!

**Reply by GaryAustin1** [reply]
Yes, you likely need to do getUser() after the updateUser.  The RLS depends on the JWT claims not what is in auth.users table.    Auth.users is never accessed by RLS as the API does not have access to it.  Things like auth.jwt() and auth.uid() are looking strictly at the jwt the client sends.

**Reply by ivanasetiawan** [reply]
Hi @GaryAustin1 ,

I have tried to run `getUser` after `updateUser` like so:

```
const { data:authData, error: authErr } = await supabase.auth.updateUser({
    data: {
        company_id,
        role: 'manager'
    }
});
const { data: { user } } = await supabase.auth.getUser();
```

and the `user` data I got back is:
```
{
...
  role: 'authenticated',
  email: 'one@one.com',
  last_sign_in_at: '2024-01-11T10:31:06.735609Z',
  app_metadata: { provider: 'email', providers: [ 'email' ] },
  user_metadata: { company_id: '37', name: 'One', role: 'manager' },
...
}
```

And then I tried both SQL expressions:
```
((((auth.jwt() -> 'user_metadata'::text) ->> 'company_id'::text))::bigint = company_id)

/* or */

auth.jwt() -> 'user_metadata' ->> 'company_id' = (company_id)::text
```

Now all users see all products as if no filter.. Am I implementing the `getUser` wrong? The data I got back can confirm that `user_metadata` has `company_id` inside it, is it still the "outdated" session?

Thank you for your patience & help @GaryAustin1! I really appreciate it!

**Reply by GaryAustin1** [reply, marked as answer]
You should try .refreshSession() instead of getUser().  I though getUser also set the session but maybe not.
Also you can test with an rpc call to a plpgsql function that has `RETURN auth.jwt()->'user_metadata::text->>'company_id';` with a text returning function.  This way you can confirm without also having the variable of RLS involved.

**Reply by GaryAustin1** [reply]
Although the data getUser has in it should be reflective of the data in auth.user independent of whether it set the session or not.  It goes to the server to get the data.   It is not clear rereading if you are saying the user data is correct from getUser or not.

**Reply by ivanasetiawan** [reply]
Thank you @GaryAustin1! It works :) 

Conclusion: I used `refreshSession` after `updateUser` and then `auth.jwt() -> 'user_metadata' ->> 'company_id' = (company_id)::text` on RLS SQL check :)

**Reply by GaryAustin1** [reply]
Thanks.  Sorry for thinking getUser did that.
